In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
digits = load_digits()
print("Done")

Done


In [8]:
print(dir(digits))

digits.keys()

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']


dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [18]:
digits.data.shape

(1797, 64)

In [13]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [16]:
digits_label = digits.target
print(digits_label.shape)
digits_label

(1797,)


array([0, 1, 2, ..., 8, 9, 8])

In [17]:
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [22]:
x = digits.data
y = digits.target
print("Done")

Done


In [24]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 5)

In [25]:
from sklearn.tree import DecisionTreeClassifier
tree_clf = DecisionTreeClassifier()
tree_clf.fit(x_train, y_train)
y_pred = tree_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.86      0.92        35
           1       0.76      0.89      0.82        35
           2       0.81      0.83      0.82        36
           3       0.97      0.78      0.87        37
           4       0.76      0.78      0.77        32
           5       0.90      0.83      0.86        46
           6       0.87      0.87      0.87        30
           7       0.86      0.86      0.86        42
           8       0.92      0.87      0.89        38
           9       0.74      1.00      0.85        29

    accuracy                           0.85       360
   macro avg       0.86      0.86      0.85       360
weighted avg       0.86      0.85      0.85       360



In [27]:
from sklearn.ensemble import RandomForestClassifier
rf_clf = RandomForestClassifier(random_state=11)
rf_clf.fit(x_train, y_train)
y_pred = rf_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        35
           1       0.97      1.00      0.99        35
           2       1.00      1.00      1.00        36
           3       1.00      0.97      0.99        37
           4       0.94      0.97      0.95        32
           5       0.98      0.96      0.97        46
           6       1.00      0.97      0.98        30
           7       0.98      0.98      0.98        42
           8       0.95      0.97      0.96        38
           9       0.93      0.97      0.95        29

    accuracy                           0.97       360
   macro avg       0.97      0.98      0.97       360
weighted avg       0.98      0.97      0.98       360



In [34]:
from sklearn.svm import SVC
svc_model = SVC()
svc_model.fit(x_train, y_train)
y_pred = svc_model.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.97      1.00      0.99        35
           2       1.00      1.00      1.00        36
           3       1.00      0.95      0.97        37
           4       1.00      1.00      1.00        32
           5       0.96      0.98      0.97        46
           6       1.00      1.00      1.00        30
           7       1.00      0.98      0.99        42
           8       0.95      0.95      0.95        38
           9       0.90      0.93      0.92        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [33]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier()
sgd_clf.fit(x_train, y_train)
y_pred = sgd_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.97      0.86      0.91        35
           2       0.92      0.97      0.95        36
           3       1.00      0.89      0.94        37
           4       1.00      0.97      0.98        32
           5       0.98      0.96      0.97        46
           6       1.00      0.97      0.98        30
           7       0.95      0.98      0.96        42
           8       0.77      0.97      0.86        38
           9       0.96      0.90      0.93        29

    accuracy                           0.95       360
   macro avg       0.96      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



In [41]:
from sklearn.linear_model import LogisticRegression
lg_reg = LogisticRegression(max_iter = 10000)
lg_reg.fit(x_train, y_train)
y_pred = lg_reg.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.89      0.97      0.93        35
           2       0.95      1.00      0.97        36
           3       0.97      0.95      0.96        37
           4       0.97      0.91      0.94        32
           5       0.98      0.93      0.96        46
           6       1.00      0.97      0.98        30
           7       0.98      0.98      0.98        42
           8       0.97      0.92      0.95        38
           9       0.88      0.97      0.92        29

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



테스트데이터 예측 결과 해석
===================

모든 라벨의 예측 결과에 대해 정밀도와 재현율이 높게 측정된 랜덤포레스트모델과 서포트벡터머신 모델이 적절한 모델이라고 예상됩니다.두 모델 모두 모든 분류에 대해 0.9 이상의 퍼포먼스를 보여주고 있고 전체 정확도도 각각 0.97과 0.98로 높습니다.

나머지 모델은 몇몇 분류에 있어 정밀도나 재현율이 낮게 나오는 경향이 있고, 정확도도 랜덤포레스트나 서포트벡터머신에 비해 낮습니다.     

랜덤포레스트 기법과 서포트벡터머신 기법 모두 우수한 결과물을 보여줬으나 해당실습에서는 최종적으로 랜덤포레스트보다 서포트벡터머신이 좀 더 좋은 선택이 되리라 생각됩니다.

랜덤포레스트는 모델에 들어갈 요인을 랜덤으로 선정하여 여러 모델을 생성한 후 앙상블하는 기법입니다. 그에 비해 서포트벡터머신은 모든 요인을 고려하여 학습한 후 경계를 만들어내어 분류하는 기법입니다. 향후에 더 양이 많고 불규칙적인 데이터에 대한 분석은 랜덤포레스트 기법이 더 맞으리라 생각되지만 해당 실습처럼 8x8의 작은 이미지의 상대적으로 규칙적인 데이터에서는 서포트벡터머신이 좀 더 논리적으로 적합한 기법이라 생각됩니다.

로직스틱리그레션의 경우 충분히 좋은 성능을 보여주고 있으나 랜덤포레스트나 서포트벡터머신에 비해 약간 아쉬운 성능을 보여주고 있습니다. 정확한 이유는 알 수 없으나 라벨의 종류가 많아 모델의 복잡도에 영향을 미쳐서 생긴 문제가 아닌가 싶습니다.